packages/react-server/src/ReactFizzServer.js

```diff
@@ -2035,29 +2061,45 @@ function flushCompletedQueues(
   request: Request,
   destination: Destination,
 ): void {
+  let allComplete = false;
   beginWriting(destination);
   try {
     // The structure of this is to go through each queue one by one and write
     // until the sink tells us to stop. When we should stop, we still finish writing
     // that item fully and then yield. At that point we remove the already completed
     // items up until the point we completed them.
 
-    // TODO: Emit preloading.
-
-    // TODO: It's kind of unfortunate to keep checking this array after we've already
-    // emitted the root.
+    let i;
     const completedRootSegment = request.completedRootSegment;
-    if (completedRootSegment !== null && request.pendingRootTasks === 0) {
-      flushSegment(request, destination, completedRootSegment);
-      request.completedRootSegment = null;
-      writeCompletedRoot(destination, request.responseState);
+    if (completedRootSegment !== null) {
+      if (request.pendingRootTasks === 0) {
+        // Flush the prelude if it exists
+        const prelude = request.prelude;
+        for (i = 0; i < prelude.length; i++) {
+          // we expect the prelude to be tiny and will ignore backpressure
+          writeChunk(destination, prelude[i]);
+        }
+        prelude.length = 0;
+
+        // Flush resources
+        writeResources(destination, request.resources);
+
+        // Flush the root segment
+        flushSegment(request, destination, completedRootSegment);
+        request.completedRootSegment = null;
+        writeCompletedRoot(destination, request.responseState);
+      } else {
+        // We haven't flushed the root yet so we don't need to check any other branches further down
+        return;
+      }
+    } else {
+      writeResources(destination, request.resources);
```

Comment:
This maybe needs to be tuned up.

At the moment, if we flush before starting work we call writeResources which happens to be empty but we don't actually know that this is safe b/c writeResources might have it's own writable content that isn't related to the number of resources that exist. For now this works by coincidence.

Typically though the flushing will happen after we've done work and thus we'll only get here if we've completed writing the root in which case we can writeResources at any time and simply want them to be near the top so the browser (or other host) receives them first